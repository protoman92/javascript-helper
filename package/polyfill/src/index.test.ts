import applyPolyfill from ".";
import { mockSomething } from "@haipham/javascript-helper-utils";

describe("Set timout polyfill", () => {
  const MAX_DELAY = Math.pow(2, 31) - 1;
  let setTimeout: jest.Mock;
  let global: typeof window;

  beforeEach(() => {
    setTimeout = jest.fn();
    global = mockSomething<typeof global>({ setTimeout: setTimeout as any });

    setTimeout.mockImplementation((handler, timeout, ...args) => {
      handler(...args, timeout);
    });

    applyPolyfill(global);
  });

  it("Should have the same behavior as normal timeout if delay is 0 or less than max", () => {
    // Setup
    const handler = jest.fn();

    // When && Then
    global.setTimeout(handler, undefined, 1, 2, 3);
    global.setTimeout(handler, NaN, 1, 2, 3);
    global.setTimeout(handler, 0, 1, 2, 3);
    global.setTimeout(handler, MAX_DELAY - 1, 1, 2, 3);

    // Then
    expect(handler.mock.calls).toMatchSnapshot();
  });

  it("Should nest timeouts if delay is more than max", () => {
    // Setup
    const handler = jest.fn();

    // When && Then
    global.setTimeout(handler, MAX_DELAY * 10 + 400, 1, 2, 3);

    // Then
    expect(handler.mock.calls).toMatchSnapshot();
  });
});
