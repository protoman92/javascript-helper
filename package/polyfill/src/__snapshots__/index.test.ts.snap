// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Set timout polyfill Should have the same behavior as normal timeout if delay is 0 or less than max 1`] = `
[
  [
    1,
    2,
    3,
    undefined,
  ],
  [
    1,
    2,
    3,
    NaN,
  ],
  [
    1,
    2,
    3,
    0,
  ],
  [
    1,
    2,
    3,
    2147483646,
  ],
]
`;

exports[`Set timout polyfill Should nest timeouts if delay is more than max 1`] = `
[
  [
    1,
    2,
    3,
    2147483647,
  ],
]
`;
